chips/rp2040/src/pio.rs

```diff
@@ -1156,68 +1372,83 @@ impl Pio {
+    pub fn handle_interrupt(&self) {
+        let ints = &self.registers.irq0_ints;
+        for (sm, irq) in self.sms.iter().zip([
+            IRQ0_INTS::SM0_TXNFULL,
+            IRQ0_INTS::SM1_TXNFULL,
+            IRQ0_INTS::SM2_TXNFULL,
+            IRQ0_INTS::SM3_TXNFULL,
+        ]) {
+            if ints.is_set(irq) {
+                sm.handle_tx_interrupt();
+            }
+        }
+        for (sm, irq) in self.sms.iter().zip([
+            IRQ0_INTS::SM0_RXNEMPTY,
+            IRQ0_INTS::SM1_RXNEMPTY,
+            IRQ0_INTS::SM2_RXNEMPTY,
+            IRQ0_INTS::SM3_RXNEMPTY,
+        ]) {
+            if ints.is_set(irq) {
+                sm.handle_rx_interrupt();
+            }
+        }
     }
 
-    /// Restart a state machine.
-    pub fn restart_sm(&self, sm_number: SMNumber) {
-        match sm_number {
-            SMNumber::SM0 => self.set_registers.ctrl.modify(CTRL::SM0_RESTART::SET),
-            SMNumber::SM1 => self.set_registers.ctrl.modify(CTRL::SM1_RESTART::SET),
-            SMNumber::SM2 => self.set_registers.ctrl.modify(CTRL::SM2_RESTART::SET),
-            SMNumber::SM3 => self.set_registers.ctrl.modify(CTRL::SM3_RESTART::SET),
+    /// Adds a program to PIO.
+    /// Call this with add_program(include_bytes!("path_to_file")).
+    pub fn add_program(&self, origin: Option<u8>, program: &[u8]) -> Result<u8, ProgramError> {
+        let mut program_u16: [u16; NUMBER_INSTR_MEMORY_LOCATIONS / 2] =
+            [0; NUMBER_INSTR_MEMORY_LOCATIONS / 2];
+        for (i, chunk) in program.chunks(2).enumerate() {
+            program_u16[i] = ((chunk[0] as u16) << 8) | (chunk[1] as u16);
         }
-    }
 
-    /// Clear a state machine’s TX and RX FIFOs.
-    pub fn sm_clear_fifos(&self, sm_number: SMNumber) {
-        self.xor_registers.sm[sm_number as usize]
-            .shiftctrl
-            .modify(SMx_SHIFTCTRL::FJOIN_RX::SET);
-        self.xor_registers.sm[sm_number as usize]
-            .shiftctrl
-            .modify(SMx_SHIFTCTRL::FJOIN_RX::SET);
+        self.add_program16(origin, &program_u16[0..program.len() / 2])
     }
 
-    /// Restart a state machine's clock divider.
-    pub fn sm_clkdiv_restart(&self, sm_number: SMNumber) {
-        match sm_number {
-            SMNumber::SM0 => self.set_registers.ctrl.modify(CTRL::CLKDIV0_RESTART::SET),
-            SMNumber::SM1 => self.set_registers.ctrl.modify(CTRL::CLKDIV1_RESTART::SET),
-            SMNumber::SM2 => self.set_registers.ctrl.modify(CTRL::CLKDIV2_RESTART::SET),
-            SMNumber::SM3 => self.set_registers.ctrl.modify(CTRL::CLKDIV3_RESTART::SET),
+    /// Adds a program to PIO.
+    /// Takes `&[u16]` as input, cause pio-asm operations are 16bit.
+    pub fn add_program16(&self, origin: Option<u8>, program: &[u16]) -> Result<u8, ProgramError> {
+        // if origin is not set, try naively to find an empty space
+        match origin {
+            Some(origin) => self
+                .try_load_program_at(origin as usize, program)
```

Comment:
why is `origin` a `u8` in this method and `add_program`, but a `usize` in `try_load_program_at`? Why does it make sense for them to be different types?

Reply:
true, it would be much easier if `origin` was `usize` everywhere